pkg/controller/master-controller-manager/application-secret-synchronizer/controller.go

```diff
@@ -0,0 +1,172 @@
+	masterManager manager.Manager,
+	seedManagers map[string]manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+) error {
+	r := &reconciler{
+		log:          log.Named(ControllerName),
+		recorder:     masterManager.GetEventRecorderFor(ControllerName),
+		masterClient: masterManager.GetClient(),
+		seedClients:  map[string]ctrlruntimeclient.Client{},
+	}
+
+	for seedName, seedManager := range seedManagers {
+		r.seedClients[seedName] = seedManager.GetClient()
+	}
+
+	c, err := controller.New(ControllerName, masterManager, controller.Options{Reconciler: r, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &k8scorev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicate.ByAnnotation(secretTypeAnnotation, "", false)); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("resource", request.Name)
+	log.Debug("Processing")
+
+	err := r.reconcile(ctx, log, request)
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+	secret := &k8scorev1.Secret{}
+
+	var err error
+	if err = r.masterClient.Get(ctx, request.NamespacedName, secret); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		} else {
+			// handling deletion
+			delSecret := &k8scorev1.Secret{ObjectMeta: v1.ObjectMeta{Name: request.Name, Namespace: seedNS}}
+			if err := r.handleDeletion(ctx, log, delSecret); err != nil {
+				return fmt.Errorf("failed to delete secret: %w", err)
+			}
+			return nil
+		}
+	}
+
+	seedsecret := secret.DeepCopy()
+	seedsecret.Namespace = seedNS
+	seedsecret.SetResourceVersion("")
+
+	namedSecretCreatorGetter := []reconciling.NamedSecretCreatorGetter{
+		secretCreator(seedsecret),
+	}
+	err = r.reconcileAllSeeds(ctx, log, seedsecret, func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
+		return reconciling.ReconcileSecrets(ctx, namedSecretCreatorGetter, seedNS, c)
+	})
+	if err != nil {
+		r.recorder.Eventf(secret, k8scorev1.EventTypeWarning, "ReconcilingError", err.Error())
+		return fmt.Errorf("reconciling secret %s failed: %w", seedsecret.Name, err)
+	}
+
+	return nil
+}
+
+func secretCreator(s *k8scorev1.Secret) reconciling.NamedSecretCreatorGetter {
+	return func() (name string, create reconciling.SecretCreator) {
+		return s.Name, func(existing *k8scorev1.Secret) (*k8scorev1.Secret, error) {
+			return s, nil
+		}
+	}
+}
+
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, secret *k8scorev1.Secret) error {
+	delfunc := func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
```

Comment:
if you want to log into this function, it's better to pass the logger as an argument, so the name of the seed is included in the log.